### PR TITLE
Instream video

### PR DESCRIPTION
**Pull request by anwzhang**

### CODE REVIEW

**Review by yoalex5 — CHANGES_REQUESTED** [review]
1. Please fix this error
```
10-28 17:16:58.846 22133-22133/org.prebid.mobile.prebidjavademo E/AndroidRuntime: FATAL EXCEPTION: main
    Process: org.prebid.mobile.prebidjavademo, PID: 22133
    java.lang.RuntimeException: Unable to pause activity {org.prebid.mobile.prebidjavademo/org.prebid.mobile.app.XandrInstreamVideoGamActivity}: java.lang.NullPointerException: Attempt to invoke virtual method 'void com.google.android.exoplayer2.SimpleExoPlayer.release()' on a null object reference
        at android.app.ActivityThread.performPauseActivity(ActivityThread.java:3381)
        at android.app.ActivityThread.performPauseActivity(ActivityThread.java:3340)
        at android.app.ActivityThread.handlePauseActivity(ActivityThread.java:3315)
        at android.app.ActivityThread.-wrap13(ActivityThread.java)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1362)
        at android.os.Handler.dispatchMessage(Handler.java:102)
        at android.os.Looper.loop(Looper.java:148)
        at android.app.ActivityThread.main(ActivityThread.java:5417)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
     Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'void com.google.android.exoplayer2.SimpleExoPlayer.release()' on a null object reference
        at org.prebid.mobile.app.XandrInstreamVideoGamActivity.releasePlayer(XandrInstreamVideoGamActivity.java:84)
        at org.prebid.mobile.app.XandrInstreamVideoGamActivity.onPause(XandrInstreamVideoGamActivity.java:142)
        at android.app.Activity.performPause(Activity.java:6348)
        at android.app.Instrumentation.callActivityOnPause(Instrumentation.java:1311)
        at android.app.ActivityThread.performPauseActivity(ActivityThread.java:3367)
        at android.app.ActivityThread.performPauseActivity(ActivityThread.java:3340) 
        at android.app.ActivityThread.handlePauseActivity(ActivityThread.java:3315) 
        at android.app.ActivityThread.-wrap13(ActivityThread.java) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1362) 
        at android.os.Handler.dispatchMessage(Handler.java:102) 
        at android.os.Looper.loop(Looper.java:148) 
        at android.app.ActivityThread.main(ActivityThread.java:5417) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
```

2. Auction bid response is empty. Should I  use some specific conditions to play an ad?
3. API_VERSION can be changed by `Build.VERSION_CODES`
4. What about MoPub AdServer ?

**Comment by anwzhang** [comment]
Hi @yoalex5 

I have fixed the crash and provided an example using stored auction response.
MoPub does not support instead video yet.
Please also check iOS implementation here: https://github.com/prebid/prebid-mobile-ios/pull/333

File: Example/PrebidDemoJava/src/main/java/org/prebid/mobile/app/RubiconInstreamVideoDemoActivity.java [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+        // parameters.setPlacement(new Signals.Placement(2));
+
+        VideoAdUnit adUnit = new VideoAdUnit("1001-1", 300, 250);
+        adUnit.setParameters(parameters);
+        this.adUnit = adUnit;
+        // Create an AdsLoader with the ad tag url.
+        adUnit.fetchDemand(new OnCompleteListener2() {
+            @Override
+            public void onComplete(ResultCode resultCode, Map<String, String> unmodifiableMap) {
+                String uri = Util.generateInstreamUriForGam("/19968336/Wei_instream_video", 640, 480, unmodifiableMap);
```
Comment (yoalex5) [inline Example/PrebidDemoJava/src/main/java/org/prebid/mobile/app/RubiconInstreamVideoDemoActivity.java:75]: Why `VideoAdUnit` has 300x250 size and when `Util.generateInstreamUriForGam` is called  640x480 is passed

File: Example/PrebidDemoJava/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -1,4 +1,7 @@
 <resources>
+    <string name="content_url"><![CDATA[https://storage.googleapis.com/gvabox/media/samples/stock.mp4]]></string>
+<!--    <string name="ad_tag_url"><![CDATA[]]></string>-->
```
Comment (yoalex5) [inline Example/PrebidDemoJava/src/main/res/values/strings.xml:3]: can we remove it ?

File: PrebidMobile/PrebidMobile-core/src/main/java/org/prebid/mobile/Util.java [hunk 3]
```diff
@@ -560,4 +559,23 @@ private static void removeUsedCustomTargetingForDFP(Object publisherAdRequest) {
...
 
+    public static String generateInstreamUriForGam(String adUnit, int w, int h, Map<String, String> keywords) {
+        String uri = "https://pubads.g.doubleclick.net/gampad/ads?";
+        if (TextUtils.isEmpty(adUnit)) {
+            throw new IllegalArgumentException("adUnit should not be empty");
+        }
+        if (w <= 0 || h <= 0) {
+            throw new IllegalArgumentException("w or h should be positive numbers");
+        }
+        uri = uri + "sz=" + w + "x" + h + "&iu=" + adUnit + "&impl=s&gdfp_req=1&env=vp&output=vast&unviewed_position_start=1";
```
Comment (yoalex5) [inline PrebidMobile/PrebidMobile-core/src/main/java/org/prebid/mobile/Util.java:570]: I have a general question regarding a required parameter: `output`

1. Does it have a correlation with `OpenRTB.Protocols`
2. [Google Docs](https://support.google.com/admanager/table/9749596?hl=en&ref_topic=2480647&visit_id=637401804158968284-34071955&rd=2) says 

> if your video player uses the IMA SDK, the output parameter for a video ad request will always be set to output=xml_vast4

And it is true if take a look at a  real request
Screenshot 2020-11-05 at 15 54 46

Should we handle it ?

File: Example/PrebidDemoJava/src/main/java/org/prebid/mobile/app/RubiconInstreamVideoDemoActivity.java (hunk 1 above)
Comment (yoalex5) [inline Example/PrebidDemoJava/src/main/java/org/prebid/mobile/app/RubiconInstreamVideoDemoActivity.java:75]: I have to create a Rubicon adUnitId and replace it

Comment (anwzhang) [reply Example/PrebidDemoJava/src/main/java/org/prebid/mobile/app/RubiconInstreamVideoDemoActivity.java:75]: Updated.

File: Example/PrebidDemoJava/src/main/res/values/strings.xml (hunk 2 above)
Comment (anwzhang) [reply Example/PrebidDemoJava/src/main/res/values/strings.xml:3]: Removed.

File: PrebidMobile/PrebidMobile-core/src/main/java/org/prebid/mobile/Util.java (hunk 3 above)
Comment (anwzhang) [reply PrebidMobile/PrebidMobile-core/src/main/java/org/prebid/mobile/Util.java:570]: Fixed.

File: Example/PrebidDemoJava/src/main/java/org/prebid/mobile/app/RubiconInstreamVideoDemoActivity.java (hunk 1 above)
Comment (anwzhang) [reply Example/PrebidDemoJava/src/main/java/org/prebid/mobile/app/RubiconInstreamVideoDemoActivity.java]: You can update the PR directly.

**Review by yoalex5 — APPROVED** [review]

**Pull request merged**